Document simplified relationships (#3590)
platformsh/platformsh-docs: sites/platform/src/create-apps/app-reference.md

## Relationships

> To allow containers in your project to communicate with each other,

Reviewer: ```suggestion
To allow containers in your project to communicate with one another,
```

you need to define relationships between them.
You can define a relationship between an app and a service, or [between two apps](/create-apps/multi-app/relationships.md).

The quickest way to define a relationship between your app and a service

> is to use the service container's default endpoint.</br>

Reviewer: "service container" might be confusing, since it also refers to a component in software architecture ([example with Symfony](https://symfony.com/doc/current/service_container.html))
```suggestion
is to use the service's default endpoint.</br>
```

However, some services allow you to define multiple databases, cores, and/or permissions.
In these cases, you can't rely on default endpoints.
Instead, you can explicitly define multiple endpoints when setting up your relationships.

{{< note >}}

> App containers don't have a default endpoint like service containers.

Reviewer: ```suggestion
App containers don't have a default endpoint like services.
```

To connect your app to another app in your project,

> explicitly define the `http` endpoint as the endpoint to connect both apps.</br>

Reviewer: ```suggestion
you need to explicitly define the `http` endpoint to connect both apps.</br>
```

For more information, see how to [define relationships between your apps](/create-apps/multi-app/relationships.md).
{{< /note >}}

To define a relationship between your app and a service:

{{< codetabs >}}

+++
title=Using default endpoints
+++

Use the following configuration:

```yaml {configFile="app"}
{{< snippet>}}
relationships:
    {{% variable "SERVICE_NAME" %}}: 
{{< /snippet >}}
```

The `SERVICE_NAME` is the name of the service as defined in its [configuration](/add-services/_index.md).

> It is used as the relationship name, and associated with a null value.

Reviewer: ```suggestion
It is used as the relationship name, and associated with a `null` value.
```

> This instructs {{% vendor/name %}} to use the service container's default endpoint to connect your app to the service.

Reviewer: ```suggestion
This instructs {{% vendor/name %}} to use the service's default endpoint to connect your app to the service.
```

For example, if you define the following configuration:

```yaml {configFile="app"}
{{< snippet>}}
relationships:
    mariadb: 
{{< /snippet>}}
```

{{% vendor/name %}} looks for a service named `mariadb` in your `{{% vendor/configfile "app" %}}` file,

> and connects your app to it through the service container's default endpoint.

Reviewer: It might worth showing the equivalent with the explicit notation:

```yaml
relationships
  mariadb:
    service: mariadb
    endpoint: mysql
```

```yaml {configFile="app"}
{{< snippet >}}
relationships:
    database: # The name of the relationship. 
        service: mariadb
        endpoint: db1
{{< /snippet >}}
```

For more information on how to handle multiple databases, multiple cores,
and/or different permissions with services that support such features,
see each service's dedicated page:

 - [MariaDB/MySQL](/add-services/mysql/_index.md#multiple-databases) (multiple databases and permissions)
 - [PostgreSQL](/add-services/postgresql/_index.md#multiple-databases) (multiple databases and permissions)
 - [Redis](/add-services/redis/_index.md#multiple-databases) (multiple databases)
 - [Solr](add-services/solr/_index.md#solr-6-and-later) (multiple cores)
 - [Vault KMS](add-services/vault/_index.md#multiple-endpoints-example) (multiple permissions)

 You can add as many relationships as you want to your app configuration:

```yaml {configFile="app"}
{{< snippet>}}
relationships:
    database1: 
        service: mariadb
        endpoint: admin
    database2:
        service: mariadb
        endpoint: legacy
    cache:
        service: redis
        endpoint: redis
    search:
        service: elasticsearch

>         endpoint: elasticsearch

Reviewer: Maybe for `cache` and `search` use the simplified notation, to show that it can be mixed with the explicit one

## Relationships

To allow containers in your project to communicate with each other,
you need to define relationships between them.
You can define a relationship between an app and a service, or [between two apps](/create-apps/multi-app/relationships.md).

You can give each relationship any name you want.
This name is then used in the [`PLATFORM_RELATIONSHIPS` environment variable](/development/variables/use-variables.md#use-provided-variables),
which provides the details of the relationship including, for example, the credentials to access the service.

{{% version/specific %}}
<!-- Platform.sh -->
To define a relationship, use one of the following syntaxes in your `{{% vendor/configfile "app" %}}` file.

<--->
<!-- Upsun -->

> To define a relationship, use one of the following syntaxes in your `{{% vendor/configfile "app" %}}` file,
> under the `applications` top-level key.

Reviewer 2: ```suggestion
To define a relationship, use one of the following syntaxes in your `{{% vendor/configfile "app" %}}` file,
under the `applications.APP_NAME` key for a given application.
```

To define a relationship between your app and a service:

{{< codetabs >}}

+++
title=Using default endpoints
+++

Use the following configuration:

```yaml {configFile="app"}
{{< snippet>}}
relationships:
    {{% variable "SERVICE_NAME" %}}: 
{{< /snippet >}}
```

The `SERVICE_NAME` is the name of the service as defined in its [configuration](/add-services/_index.md).
It is used as the relationship name, and associated with a `null` value.
This instructs {{% vendor/name %}} to use the service's default endpoint to connect your app to the service.

For example, if you define the following configuration:

```yaml {configFile="app"}
{{< snippet>}}
relationships:
    mariadb: 
{{< /snippet>}}
```

> {{% vendor/name %}} looks for a service named `mariadb` in your `{{% vendor/configfile "app" %}}` file,

Author: Good catch! Fixed

{{< note >}}

> To connect your app to another app in your project, replace the `SERVICE_NAME` by the name of that other app,

Reviewer 2: Given

> Every service container has a default endpoint that is used if no endpoint is defined.

Is it smart enough to know that `endpoint: http` is the default to be applied?

If so, maybe we can include

```yaml {configFile="app"}
relationships:
    api:
        service: backend
```

And _actually_ is it ever necessary that a user knows about `http` in the background if this is the way we're handling it now?

Author: My assumption was bad actually, app containers don't have a default `http` endpoint. Updated accordingly.

and use the `http` endpoint.
For more information, see how to [define relationships between your apps](/create-apps/multi-app/relationships.md).
{{< /note >}}

You can add as many relationships as you want to your app configuration:

```yaml {configFile="app"}
relationships:

>   database1: 
>     service: mariadb
>     endpoint: mysql
>   database2:
>     service: mariadb
>     endpoint: mysql
>   cache:
>     service: redis
>     endpoint: redis
>   search:
>     service: elasticsearch
>     endpoint: elasticsearch

Reviewer 2: ```suggestion
relationships:
    database1: 
        service: mariadb
        endpoint: admin
    database2:
        service: mariadb
        endpoint: legacy
    api:
        service: backend        
    cache:
        service: redis
    search:
        service: elasticsearch
```

You can add as many relationships as you want to your app configuration:

```yaml {configFile="app"}
{{< snippet>}}
relationships:
    mariadb:  
    redis: 
    elasticsearch: 
{{< /snippet>}}
```

<--->

+++
title=Using explicit endpoints
+++

Use the following configuration:

```yaml {configFile="app"}
{{< snippet >}}
relationships:
    {{% variable "RELATIONSHIP_NAME" %}}:
        service: {{% variable "SERVICE_NAME" %}} 
        endpoint: {{% variable "ENDPOINT_NAME" %}}
{{< /snippet >}}
```
 
- `RELATIONSHIP_NAME` is the name you want to give to the relationship.
- `SERVICE_NAME` is the name of the service as defined in its [configuration](/add-services/_index.md).

> - `ENDPOINT_NAME` is the endpoint to connect to the service.

Reviewer: ```suggestion
- `ENDPOINT_NAME` is the endpoint your app will use to connect to the service (refer to the service reference to know which value to use).
```

and/or different permissions with services that support such features,
see each service's dedicated page:

 - [MariaDB/MySQL](/add-services/mysql/_index.md#multiple-databases) (multiple databases and permissions)
 - [PostgreSQL](/add-services/postgresql/_index.md#multiple-databases) (multiple databases and permissions)
 - [Redis](/add-services/redis/_index.md#multiple-databases) (multiple databases)
 - [Solr](add-services/solr/_index.md#solr-6-and-later) (multiple cores)
 - [Vault KMS](add-services/vault/_index.md#multiple-endpoints-example) (multiple permissions)

 You can add as many relationships as you want to your app configuration,
 using both default and explicit endpoints according to your needs:

```yaml {configFile="app"}
{{< snippet>}}
relationships:
    database1: 
        service: mariadb
        endpoint: admin
    database2:
        service: mariadb
        endpoint: legacy
    cache:
        redis:
    search:
        elasticsearch:
{{< /snippet>}}
```

{{< /codetabs >}}

{{< version/only "1" >}}

{{< note >}}

The following legacy syntax for specifying relationships is still supported by {{% vendor/name %}}:

```yaml

> relationships

Reviewer 2: ```suggestion
relationships:
```

    <RELATIONSHIP_NAME>: "<SERVICE_NAME>:<ENDPOINT_NAME>"
```

For example:

```yaml

> relationships

Reviewer 2: ```suggestion
relationships:
```
